src/m365/spo/commands/file/file-rename.ts

```diff
@@ -0,0 +1,161 @@
+  }
+
+  public getTelemetryProperties(args: CommandArgs): any {
+    const telemetryProps: any = super.getTelemetryProperties(args);
+    telemetryProps.force = !!args.options.force;
+    return telemetryProps;
+  }
+
+  public commandAction(logger: Logger, args: CommandArgs, cb: () => void): void {
+    const webUrl = args.options.webUrl;
+    const originalFileServerRelativeUrl: string = urlUtil.getServerRelativePath(args.options.webUrl, args.options.sourceUrl);
+    // Check if the source file exists.
+    // Called on purpose, we explicitly check if user specified file exists, otherwise, the rename can't happen and we will throw an error
+    this
+      .fileExists(originalFileServerRelativeUrl, webUrl)
+      .then((): Promise<void> => {
+        // Check if file already exists with the new name, if so: the file will be removed to the recycle bin
+        if (args.options.force) {
+          // delete the target file if force is set
+          const targetFileServerRelativeUrl: string = `${urlUtil.getServerRelativePath(webUrl, args.options.sourceUrl.substring(0, args.options.sourceUrl.lastIndexOf('/')))}/${args.options.targetFileName}`;
+          return this.recycleFile(webUrl, targetFileServerRelativeUrl, logger); 
+        }
+
+        return Promise.resolve();
+      })
+      .then((): Promise<void> => {
+        // all preconditions met, now rename item
+        const requestBody: any = {
+          formValues : [{
+            FieldName: 'FileLeafRef',
+            FieldValue: args.options.targetFileName
+          }]
+        };
+
+        const requestOptions: any = {
+          url: `${webUrl}/_api/web/GetFileByServerRelativeUrl('${encodeURIComponent(originalFileServerRelativeUrl)}')/ListItemAllFields/ValidateUpdateListItem()`,
+          headers: {
+            'accept': 'application/json;odata=nometadata'
+          },
+          data: requestBody,
+          responseType: 'json'
+        };
+
+        return request.post(requestOptions);
+      })
+      .then((resp: any): Promise<any> => {
+        return new Promise<void>((resolve: () => void): void => {
+          logger.log(resp.value);
+          resolve();
+        });
+      })
+      .then(_ => cb(), (err: any): void => this.handleRejectedODataJsonPromise(err, logger, cb));
+  }
+
+  /**
+   * Checks if the original file exists
+   */
+  private fileExists(originalFileServerRelativeUrl: string, webUrl: string): Promise<void> {
+    const requestUrl = `${webUrl}/_api/web/GetFileByServerRelativeUrl('${encodeURIComponent(originalFileServerRelativeUrl)}')`;
+    const requestOptions: any = {
+      url: requestUrl,
+      method: 'GET',
+      headers: {
+        'accept': 'application/json;odata=nometadata'
+      },
+      responseType: 'json'
+    };
+    return request.get(requestOptions);
+  }
+
+  /**
+   * Deletes file in the site recycle bin
+   */
+  private recycleFile(webUrl: string, targetFileServerRelUrl: string, logger: Logger): Promise<void> {
```

Comment:
I wonder if we don't 'reinvent the wheel' here a bit 🤔 I mean we already have a file remove command which has the exact same logic and may recycle the file. So I think we may consider not rewriting it but reuse the logic using `Cli.executeCommandWithOutput`

@pnp/cli-for-microsoft-365-maintainers  what is your opinion on this case 🤔